synapse/logging/tracing.py

```diff
@@ -0,0 +1,1022 @@
+
+
+def create_non_recording_span() -> Optional["opentelemetry.trace.Span"]:
+    """Create a no-op span that does not record or become part of a recorded trace"""
+    if opentelemetry is None:
+        return None  # type: ignore[unreachable]
+
+    return opentelemetry.trace.NonRecordingSpan(
+        opentelemetry.trace.INVALID_SPAN_CONTEXT
+    )
+
+
+def start_span(
+    name: str,
+    *,
+    context: Optional["opentelemetry.context.context.Context"] = None,
+    kind: Optional["opentelemetry.trace.SpanKind"] = SpanKind.INTERNAL,
+    attributes: "opentelemetry.util.types.Attributes" = None,
+    links: Optional[Sequence["opentelemetry.trace.Link"]] = None,
+    start_time: Optional[int] = None,
+    record_exception: bool = True,
+    set_status_on_exception: bool = True,
+    end_on_exit: bool = True,
+    # For testing only
+    tracer: Optional["opentelemetry.trace.Tracer"] = None,
+) -> "opentelemetry.trace.Span":
+    if opentelemetry is None:
+        raise Exception("Not able to create span without opentelemetry installed.")
+
+    if tracer is None:
+        tracer = opentelemetry.trace.get_tracer(__name__)
+
+    # TODO: Why is this necessary to satisfy this error? It has a default?
+    #  ` error: Argument "kind" to "start_span" of "Tracer" has incompatible type "Optional[SpanKind]"; expected "SpanKind"  [arg-type]`
+    if kind is None:
+        kind = SpanKind.INTERNAL
+
+    return tracer.start_span(
+        name=name,
+        context=context,
+        kind=kind,
+        attributes=attributes,
+        links=links,
+        start_time=start_time,
+        record_exception=record_exception,
+        set_status_on_exception=set_status_on_exception,
+    )
+
+
+def start_active_span(
+    name: str,
+    *,
+    context: Optional["opentelemetry.context.context.Context"] = None,
+    kind: "opentelemetry.trace.SpanKind" = SpanKind.INTERNAL,
+    attributes: "opentelemetry.util.types.Attributes" = None,
+    links: Optional[Sequence["opentelemetry.trace.Link"]] = None,
+    start_time: Optional[int] = None,
+    record_exception: bool = True,
+    set_status_on_exception: bool = True,
+    end_on_exit: bool = True,
+    # For testing only
+    tracer: Optional["opentelemetry.trace.Tracer"] = None,
+) -> ContextManager[Optional["opentelemetry.trace.Span"]]:
+    if opentelemetry is None:
+        return contextlib.nullcontext()  # type: ignore[unreachable]
+
+    span = start_span(
+        name=name,
+        context=context,
+        kind=kind,
+        attributes=attributes,
+        links=links,
+        start_time=start_time,
+        record_exception=record_exception,
+        set_status_on_exception=set_status_on_exception,
+        tracer=tracer,
+    )
+
+    # Equivalent to `tracer.start_as_current_span`
+    return opentelemetry.trace.use_span(
+        span,
+        end_on_exit=end_on_exit,
+        record_exception=record_exception,
+        set_status_on_exception=set_status_on_exception,
+    )
```

Comment:
*Moving this over from https://github.com/matrix-org/synapse/pull/13440#discussion_r940725431,*

Occasionally, I see spans that have a super massive duration and shows up with `213503982d 8h` in the Jaeger UI. Obviously looks like some sort of max value bug.

<img width="1194" alt="" src="https://user-images.githubusercontent.com/558581/183526530-66c391a8-0fe6-4085-84d6-db24621c3469.png">


The raw data for those spans are the following. Only these spans are messed up and the `end_time` does not affect any more parent spans before `_process_pulled_event`.

`db.get_partial_state_events`
```jsonc
{
    "traceID": "563478e5d75db03e9fc028822fde0649",
    "spanID": "5b63b0172763c5cf",
    "flags": 1,
    "operationName": "db.get_partial_state_events",
    "references": [
        {
            "refType": "CHILD_OF",
            "traceID": "563478e5d75db03e9fc028822fde0649",
            "spanID": "10c9f2a2c50e286f"
        }
    ],
    "startTime": 1659994809436162,
    "duration": 18446744073709517944,
    "tags": [
        // ...
    ],
    "logs": [],
    "processID": "p1",
    "warnings": null
}
```

`_process_pulled_event`
```
{
    "traceID": "563478e5d75db03e9fc028822fde0649",
    "spanID": "6e9ee2608b03b542",
    "flags": 1,
    "operationName": "_process_pulled_event",
    "references": [
        {
            "refType": "CHILD_OF",
            "traceID": "563478e5d75db03e9fc028822fde0649",
            "spanID": "af2c9c8d08284134"
        }
    ],
    "startTime": 1659994809432356,
    "duration": 18446744073709532019,
    "tags": [
        // ...
    ],
    "logs": [],
    "processID": "p1",
    "warnings": null
},
```

Definitely close to the Python "max int size" (unsigned word, 64 bits): `sys.maxsize * 2 + 1` (or `2^64 - 1`) -> `18446744073709551615` (via https://stackoverflow.com/questions/7604966/maximum-and-minimum-values-for-ints)

It seems like it's some negative value getting turned into an unsigned integer so it wraps to the end of the value range. But I don't see the math where that can happen yet.